### PR TITLE
[Dev] Run correct cookbook version based on branch and deploy development cookbooks and stable cookbooks

### PR DESCRIPTION
**Pull request by raulcd**
This PR makes some changes on the current GitHub actions workflows to work with both the current development version of Arrow and the stable version.

On commits on the default branch (main) the development cookbooks will be deployed. I've tested on my branch setting it up on GitHub pages: https://raul.dev/arrow-cookbook/dev/
As can be seen there (on the Java or Python cookbooks) the version used is the nightlies one: https://raul.dev/arrow-cookbook/dev/java/index.html or https://raul.dev/arrow-cookbook/dev/py/index.html
Example of workflow run: https://github.com/raulcd/arrow-cookbook/actions/runs/3731385491

On commits on the stable branch the stable cookbooks will be deployed: https://raul.dev/arrow-cookbook/index.html with the current stable version being used to build: https://raul.dev/arrow-cookbook/java/index.html or https://raul.dev/arrow-cookbook/py/index.html
Example of workflow run: https://github.com/raulcd/arrow-cookbook/actions/runs/3731189658

On PRs against main the nightlies version will be used, example of PR: https://github.com/raulcd/arrow-cookbook/pull/11/checks
Using the nightlies version: https://github.com/raulcd/arrow-cookbook/actions/runs/3731199253/jobs/6329147897#step:4:88

On PRs against stable the stable version will be used. Example of PR: https://github.com/raulcd/arrow-cookbook/pull/12/checks
Using the stable version: https://github.com/raulcd/arrow-cookbook/actions/runs/3732065139/jobs/6331024764#step:4:74

### CODE REVIEW

File: .github/workflows/test_java_cookbook.yml [hunk 1]
```diff
@@ -54,6 +52,51 @@ jobs:
...
+            export ARROW_NIGHTLY=0
+          fi
+          make javatest
+      - name: Build cookbook
+        run: |
+          if [ "${{ github.base_ref }}" == "main" ]; then
+            export ARROW_NIGHTLY=1
+          else
+            export ARROW_NIGHTLY=0
+          fi
```
Comment (assignUser) [inline .github/workflows/test_java_cookbook.yml:68]: You could move this to to env using gha expressions:
```yaml
env:
  ARROW_R_NIGHTLY: ${{ github.base_ref == 'main' && 1 || 0 }}
```

Comment (assignUser) [reply .github/workflows/test_java_cookbook.yml:68]: (same for the other occurences ofc) the general foirm is `${{ <condition> && <string_true> || <string_false> }}`

**Review by assignUser — APPROVED** [review]
some minor comments but nothing blocking :+1:  

In a follow up we could refactor the setup part that happens in multiple workflows into a local action but that is polish and not strictly necessary :)

File: .github/workflows/deploy_development_cookbooks.yml [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+        uses: actions/cache@v2
+        env:
+          # Increase this value to reset cache if cpp/dev.yml has not changed
+          CACHE_NUMBER: 0
+        with:
+          path: ~/conda_pkgs_dir
+          key:
+            ${{ runner.os }}-conda-${{ env.CACHE_NUMBER }}-${{ hashFiles('cpp/dev.yml') }}
+      - name: Setup environment
+        uses: conda-incubator/setup-miniconda@v2
```
[the remark this answers is not in the document]
Comment (raulcd) [reply .github/workflows/deploy_development_cookbooks.yml:64]: I did not know there was a list of allowed actions. Where can I find that list? From what I can see here: https://infra.apache.org/github-actions-policy.html we should pin it to a specific commit hash (same as all the r-lib ones). We can also cover this on a different issue as I am not adding new actions but using the ones we are already using. In this case this one is already used to deploy the cookbooks today.

[the remark this answers is not in the document]
Comment (assignUser) [reply .github/workflows/deploy_development_cookbooks.yml:66]: Ah ok :+1:  Yes there is an allow-regex somewhere but I am not sure where I found that...

**Comment by assignUser** [comment]
+1

**Pull request merged**